File: ios/Classes/FlutterBlePeripheralPlugin.swift
import Flutter
import UIKit
import CoreBluetooth

public class FlutterBlePeripheralPlugin: NSObject, FlutterPlugin {
    private var peripheralManager: PeripheralManager?
    private let channel: FlutterMethodChannel
    
    public init(channel: FlutterMethodChannel) {
        self.channel = channel
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: FlutterBlePeripheralConstants.methodChannel, binaryMessenger: registrar.messenger())
        let instance = FlutterBlePeripheralPlugin(channel: channel)
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case FlutterBlePeripheralConstants.initialize:
            peripheralManager = PeripheralManager()
            result(nil)
        case FlutterBlePeripheralConstants.startAdvertising:
            if let args = call.arguments as? [String: Any] {
                peripheralManager?.startAdvertising(args: args)
                result(nil)
            } else {
                result(FlutterError.init(code: FlutterBlePeripheralError.invalidArguments.rawValue, message: "Invalid arguments", details: nil))
            }
        case FlutterBlePeripheralConstants.stopAdvertising:
            peripheralManager?.stopAdvertising()
            result(nil)
        case FlutterBlePeripheralConstants.addService:
            if let args = call.arguments as? [String: Any], let service = BleService.fromMap(args) {
                peripheralManager?.addService(service: service)
                result(nil)
            } else {
                result(FlutterError.init(code: FlutterBlePeripheralError.invalidArguments.rawValue, message: "Invalid arguments", details: nil))
            }
        case FlutterBlePeripheralConstants.sendData:
            if let args = call.arguments as? [String: Any], let uuidString = args["characteristicUUID"] as? String, let data = args["data"] as? [UInt8] {
                let uuid = CBUUID(string: uuidString)
                peripheralManager?.sendData(characteristicUUID: uuid, data: Data(data))
                result(nil)
            } else {
                result(FlutterError.init(code: FlutterBlePeripheralError.invalidArguments.rawValue, message: "Invalid arguments", details: nil))
            }
        case FlutterBlePeripheralConstants.setCharacteristicValueUpdateCallback:
            if let args = call.arguments as? [String: Any], let uuidStrings = args["characteristicUUIDs"] as? [String] {
               let uuids = uuidStrings.map { CBUUID(string: $0) }
                    peripheralManager?.setCharacteristicValueUpdateCallback { characteristic, data in
                        if uuids.contains(characteristic.uuid) {
                            let args: [String: Any] = [
                                "characteristicUUID": characteristic.uuid.uuidString,
                                "value": data
                            ]
                            self.channel.invokeMethod("onCharacteristicUpdate", arguments: args)
                        }
                    }
                    result(nil)
                } else {
                    result(FlutterError.init(code: FlutterBlePeripheralError.invalidArguments.rawValue, message: "Invalid arguments", details: nil))
                }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
//    private func setCharacteristicValueUpdateCallback(call: FlutterMethodCall, result: @escaping FlutterResult) {
//        guard let args = call.arguments as? [String: Any],
//              let characteristicUUIDString = args["characteristicUUID"] as? String else {
//            result(FlutterError(code: "invalid_argument", message: "Invalid arguments", details: nil))
//            return
//        }
//        guard let peripheralManager = peripheralManager else {
//            result(FlutterError(code: "peripheral_manager_unavailable", message: "Peripheral manager unavailable", details: nil))
//            return
//        }
//        let characteristicUUID = CBUUID(string: characteristicUUIDString)
//        let valueUpdateCallback = { (peripheral: CBPeripheral, characteristic: CBCharacteristic) in
//            let value = characteristic.value ?? Data()
//            let arguments: [String: Any] = [
//                "characteristicUUID": characteristic.uuid.uuidString,
//                "value": [UInt8](value)
//            ]
//            self.eventSink?("onCharacteristicUpdate", arguments)
//        }
//        valueUpdateCallbacks[characteristicUUID] = valueUpdateCallback
//        result(nil)
//    }

}

File: android/src/main/kotlin/com/regal/flutter_ble_peripheral/FlutterBlePeripheralPlugin.kt
package com.regal.flutter_ble_peripheral

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class FlutterBlePeripheralPlugin : FlutterPlugin, MethodCallHandler {
  private lateinit var channel: MethodChannel
  private lateinit var peripheralManager: PeripheralManager

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, FlutterBlePeripheralConstants.methodChannel)
    channel.setMethodCallHandler(this)
    peripheralManager = PeripheralManager(flutterPluginBinding.applicationContext)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      FlutterBlePeripheralConstants.initialize -> {
        peripheralManager.initialize()
        result.success(null)
      }
      FlutterBlePeripheralConstants.startAdvertising -> {
        val advertisementData = call.arguments as? Map<String, Any>
        peripheralManager.startAdvertising(advertisementData)
        result.success(null)
      }
      FlutterBlePeripheralConstants.stopAdvertising -> {
        peripheralManager.stopAdvertising()
        result.success(null)
      }
      FlutterBlePeripheralConstants.addService -> {
        val serviceData = call.arguments as? Map<String, Any>
        if (serviceData != null) {
          val service = BleService.fromMap(serviceData)
          if (service != null) {
            peripheralManager.addService(service)
            result.success(null)
          } else {
            result.error("INVALID_ARGUMENTS", "Invalid service data", null)
          }
        } else {
          result.error("INVALID_ARGUMENTS", "Invalid arguments", null)
        }
      }
      FlutterBlePeripheralConstants.sendData -> {
        val arguments = call.arguments as? Map<String, Any>
        val characteristicUUID = arguments?.get("characteristicUUID") as? String
        val data = arguments?.get("data") as? List<Int>
        if (characteristicUUID != null && data != null) {
          peripheralManager.sendData(characteristicUUID, data)
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENTS", "Invalid arguments", null)
        }
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}
